lib/srv/db/sqlserver/kinit/kinit_test.go

```diff
@@ -0,0 +1,133 @@
+package kinit
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+//go:embed kinit.cache
+var cache []byte
+
+type staticCache struct {
+	t    *testing.T
+	pass bool
+}
+
+func (s *staticCache) CommandContext(ctx context.Context, name string, args ...string) *exec.Cmd {
+	cachePath := args[len(args)-1]
+	require.NotEmpty(s.t, cachePath)
+	err := os.WriteFile(cachePath, cache, 0664)
+	require.NoError(s.t, err)
+
+	if s.pass {
+		return exec.Command("echo")
+	}
+	cmd := exec.Command("")
+	cmd.Err = errors.New("bad command")
+	return cmd
+
+}
+
+type testCertGetter struct {
+	pass bool
+}
+
+func (t *testCertGetter) GetCertificateBytes(context.Context) ([]byte, []byte, []byte, error) {
+	if t.pass {
+		return nil, nil, nil, nil
+	}
+	return nil, nil, nil, errors.New("could not get cert bytes")
+
+}
+
+type testStep struct {
+	initializer *PKInit
+	expectErr   func(t require.TestingT, err error, msgAndArgs ...interface{})
+	expectNil   func(t require.TestingT, object interface{}, msgAndArgs ...interface{})
+	post        func() error
+}
+
+func step(s *staticCache, c *testCertGetter, expectErr func(t require.TestingT, err error, msgAndArgs ...interface{}), expectNil func(t require.TestingT, object interface{}, msgAndArgs ...interface{})) *testStep {
+	dir, err := os.MkdirTemp("", "krb5_cache")
+	require.NoError(s.t, err)
+
+	return &testStep{
+		initializer: New(NewCommandLineInitializerWithCommand(nil,
+			"alice",
+			"example.com",
+			"host.example.com",
+			"host.example.com",
+			dir, nil, s, c)),
+		expectErr: expectErr,
+		expectNil: expectNil,
+		post: func() error {
+			return os.RemoveAll(dir)
+		},
+	}
+}
+
+func TestNewWithCommandLineProvider(t *testing.T) {
+
+	steps := []*testStep{
```

Comment:
Are these more like steps or test cases ? They look more like individual cases, if so, can we name them more like that and also use `t.Run()` when executing them to break up the results a bit more. This makes it clearer which case is causing the failure.